zilliqa/src/exec.rs

```diff
@@ -826,14 +890,183 @@ impl PendingState {
-                Ok(vac.insert(account))
+    pub fn load_account(&mut self, address: Address) -> Result<&mut PendingAccount> {
+        load_account(&self.pre_state, &mut self.new_state, address)
+    }
+
+    pub fn load_var_info(&mut self, address: Address, variable: &str) -> Result<(&str, u8)> {
+        let account = self.load_account(address)?;
+        let Code::Scilla { types, .. } = &account.account.code else {
+            return Err(anyhow!("not a scilla contract"));
+        };
+        let (ty, depth) = types
+            .get(variable)
+            .ok_or_else(|| anyhow!("missing type for variable: {variable}"))?;
+        Ok((ty, *depth))
+    }
+
+    /// Set the value of a given variable at the given indices. This can be used for setting map values, unlike
+    /// `load_storage`.
+    pub fn set_storage(
+        &mut self,
+        address: Address,
+        var_name: &str,
+        indices: &[Vec<u8>],
+        value: StorageValue,
+    ) -> Result<()> {
+        let account = load_account(&self.pre_state, &mut self.new_state, address)?;
+
+        let mut current = account
+            .storage
+            .entry(var_name.to_owned())
+            .or_insert_with(StorageValue::incomplete);
+        for key in indices {
+            let current_map = match current {
+                StorageValue::Map { map, .. } => map,
+                StorageValue::Value(Some(_)) => {
+                    return Err(anyhow!("expected a map"));
+                }
+                StorageValue::Value(None) => {
+                    unimplemented!("deletes of whole maps are unsupported")
+                }
+            };
+            let child = current_map
+                .entry(key.clone())
+                .or_insert_with(StorageValue::incomplete);
+            current = child;
+        }
+
+        *current = value;
+
+        Ok(())
+    }
+
+    pub fn load_storage(
+        &mut self,
+        address: Address,
+        var_name: &str,
+        indices: &[Vec<u8>],
+    ) -> Result<&mut Option<Bytes>> {
+        let account = load_account(&self.pre_state, &mut self.new_state, address)?;
+
+        fn get_cached<'a>(
+            storage: &'a mut BTreeMap<String, StorageValue>,
+            var_name: &str,
+            indices: &[Vec<u8>],
+        ) -> Result<(&'a mut StorageValue, bool)> {
+            let mut cached = true;
+            let mut current = storage.entry(var_name.to_owned()).or_insert_with(|| {
+                cached = false;
+                StorageValue::incomplete()
+            });
+            for key in indices {
+                let current_map = match current {
+                    StorageValue::Map { map, .. } => map,
+                    StorageValue::Value(Some(_)) => {
+                        return Err(anyhow!("expected a map"));
+                    }
+                    StorageValue::Value(None) => {
+                        unimplemented!("deletes of whole maps are unsupported")
```

Comment:
Whilst they may be, that's not what this unimplemented! should be saying :-p

Reply:
See above.

Reply:
Agree.  Perhaps worth a:

```
// StorageValue::Value(None) means delete, and since we don't support deleting whole elements here .. 
```
 ?